libavcodec/vaapi_encode.c

```diff
@@ -1177,19 +1180,46 @@ int ff_vaapi_encode_receive_packet(AVCodecContext *avctx, AVPacket *pkt)
             return AVERROR(EAGAIN);
     }
 
-    pic = NULL;
-    err = vaapi_encode_pick_next(avctx, &pic);
-    if (err < 0)
-        return err;
-    av_assert0(pic);
+    while (av_fifo_size(ctx->encode_fifo) <
+           (ctx->async_depth > MAX_PICTURE_REFERENCES ?
+            ctx->async_depth : (MAX_PICTURE_REFERENCES + 1)) *
+            sizeof(VAAPIEncodePicture *)) {
+        pic = NULL;
+        err = vaapi_encode_pick_next(avctx, &pic);
+        if (err < 0)
+            break;
+        av_assert0(pic);
 
-    pic->encode_order = ctx->encode_order++;
+        pic->encode_order = ctx->encode_order +
+                            (av_fifo_size(ctx->encode_fifo) / sizeof(VAAPIEncodePicture *));
 
-    err = vaapi_encode_issue(avctx, pic);
-    if (err < 0) {
-        av_log(avctx, AV_LOG_ERROR, "Encode failed: %d.\n", err);
+        err = vaapi_encode_issue(avctx, pic);
+        if (err < 0) {
+            av_log(avctx, AV_LOG_ERROR, "Encode failed: %d.\n", err);
+            return err;
+        }
+
+        av_fifo_generic_write(ctx->encode_fifo, &pic, sizeof(pic), NULL);
+    }
+
+    if (!av_fifo_size(ctx->encode_fifo))
         return err;
+
+#if VA_CHECK_VERSION(1, 9, 0)
+    if (av_fifo_size(ctx->encode_fifo) < ctx->async_depth * sizeof(VAAPIEncodePicture *) &&
+        !ctx->end_of_stream) {
+        av_fifo_generic_peek(ctx->encode_fifo, &pic, sizeof(pic), NULL);
+        err = vaSyncBuffer(ctx->hwctx->display, pic->output_buffer, 0);
```

Comment:
we'd better output all ready buffers. not only the current buffer, it will reduce the delay